### PR TITLE
[ML] Stop cross-validation early if the parameters have high predicted test loss

### PR DESCRIPTION
**Pull request by tveasey**
This makes two changes:
1. It sets the number of folds to use based on the amount of data available: we ensure each training fold has sufficient data per feature.
2. During hyperparameter optimisation, we use a linear model to predict the cross-validation loss on the remaining folds and stop early if the predicted test loss has a low chance of being less than the best values found so far.

### CODE REVIEW

File: lib/maths/CBoostedTreeImpl.cc [hunk 1]
```diff
@@ -992,6 +1044,114 @@ CBoostedTreeImpl::trainTree(core::CDataFrame& frame,
...
+    std::sort(ordered.begin(), ordered.end());
+    CSetTools::inplace_set_difference(present, ordered.begin(), ordered.end());
+    LOG_TRACE(<< "present = " << core::CContainerPrinter::print(present));
+
+    // Get the current round feature vector.
+    TVector x(2 * present.size());
+    for (std::size_t col = 0; col < present.size(); ++col) {
+        x(col) = *m_FoldRoundTestLosses[present[col]][m_CurrentRound];
+        x(present.size() + col) = 0.0;
+    }
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: Maybe, you can move this part of the code to where `x` is actually used.

File: lib/maths/CBoostedTreeImpl.cc [hunk 2]
```diff
@@ -992,6 +1044,114 @@ CBoostedTreeImpl::trainTree(core::CDataFrame& frame,
...
+
+    // We have a subset of folds for which we've computed test loss. We want to
+    // estimate the test loss we'll see for the remaining folds to decide if it
+    // is worthwhile to continue training with these parameters and to correct
+    // the loss value supplied to Bayesian Optimisation to account for the folds
+    // we haven't trained on. We achieve this by for each missing fold fitting an
+    // OLS to the data (x_i, loss(m_i)) where i ranges over the previous rounds
+    // and x_i is the i'th vector whose components comprise the losses for which
+    // we have values in the current round and indicators for whether they were
+    // missing in the i'th round. We only include a round if we've trained for at
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc]: Also, what is `loss(m_i)`? Judging from the code below, it's some sort of indicator variable.

File: lib/maths/CBoostedTreeImpl.cc [hunk 3]
```diff
@@ -992,6 +1044,114 @@ CBoostedTreeImpl::trainTree(core::CDataFrame& frame,
...
+
+        TMeanVarAccumulator residualMoments;
+        for (int row = 0; row < A.rows(); ++row) {
+            residualMoments.add(b(row) - A.row(row) * params);
+        }
+
+        double predictedTestLoss{params.transpose() * x};
+        double predictedTestLossVariance{
+            CBasicStatistics::maximumLikelihoodVariance(residualMoments)};
+        predictedTestLosses.push_back(CBasicStatistics::momentsAccumulator(
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeImpl.cc]: IMO: I would move this line further below, it's kind of getting lost here and then I am wondering how do you feel the variable `predictedTestLosses` you are returning at the end of the function.

File: lib/maths/CBoostedTreeFactory.cc [hunk 4]
```diff
@@ -229,6 +230,46 @@ void CBoostedTreeFactory::initializeMissingFeatureMasks(const core::CDataFrame&
...
+                },
+                std::size_t{0}));
+        std::size_t totalNumberTrainingRows{0};
+        for (const auto& numberTrainingRows : result.first) {
+            totalNumberTrainingRows += numberTrainingRows.s_FunctionState;
+        }
+        LOG_TRACE(<< "total number training rows = " << totalNumberTrainingRows);
+
+        // We require at least twice the number of rows we'll sample in a bag per
+        // fold if possible. In order to estimate this we use the number of input
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: I've expanded this comment in [01c5d45](https://github.com/elastic/ml-cpp/pull/915/commits/01c5d4558a01dc051edaec8bf44fce2d2b29c721).

File: lib/maths/CBoostedTreeFactory.cc [hunk 5]
```diff
@@ -229,6 +230,46 @@ void CBoostedTreeFactory::initializeMissingFeatureMasks(const core::CDataFrame&
...
+        // features as a proxy for the number of features we'll actually use after
+        // feature selection.
+        double desiredTrainingFraction{(m_InitialDownsampleRowsPerFeature *
+                                        static_cast<double>(frame.numberColumns() - 1)) /
+                                       static_cast<double>(totalNumberTrainingRows)};
+        if (2.0 * desiredTrainingFraction >= 1.0 - 1.0 / static_cast<double>(MAX_NUMBER_FOLDS)) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: I've rejigged in [294e0b9](https://github.com/elastic/ml-cpp/pull/915/commits/294e0b92d05a3769d4aef17b9b00f1089634d49d).

File: lib/maths/CBoostedTreeFactory.cc [hunk 6]
```diff
@@ -229,6 +230,46 @@ void CBoostedTreeFactory::initializeMissingFeatureMasks(const core::CDataFrame&
...
+            m_TreeImpl->m_NumberFolds = MAX_NUMBER_FOLDS;
+        } else {
+            m_TreeImpl->m_NumberFolds = static_cast<std::size_t>(
+                std::ceil(1.0 / (1.0 - 2.0 * desiredTrainingFraction)));
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeFactory.cc]: See comment added in [01c5d45](https://github.com/elastic/ml-cpp/pull/915/commits/01c5d4558a01dc051edaec8bf44fce2d2b29c721).

File: lib/maths/CBoostedTreeImpl.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: This wasn't well written. I've expanded along the lines you suggest in [b6943e4](https://github.com/elastic/ml-cpp/pull/915/commits/b6943e491d3c3c85e480703ff2bc0963ace40527).

File: lib/maths/CBoostedTreeImpl.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: This is computed outside the loop over missing folds since it is fixed for all folds. It probably doesn't really matter but it would be less efficient to push into the loop. I've added a comment to this effect, WDYT?

File: lib/maths/CBoostedTreeImpl.cc (hunk 3 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc]: See [de99d29](https://github.com/elastic/ml-cpp/pull/915/commits/de99d2987b75ef2a83fe21ab0084fa69786d93d9).

**Comment by tveasey** [comment]
Thanks for the review @valeriy42. I think I've now addressed all your comments.

File: lib/maths/CBoostedTreeFactory.cc [hunk 7]
```diff
@@ -46,7 +46,10 @@ const double MIN_DOWNSAMPLE_LINE_SEARCH_RANGE{2.0};
 const double MAX_DOWNSAMPLE_LINE_SEARCH_RANGE{144.0};
 const double MIN_DOWNSAMPLE_FACTOR_SCALE{0.3};
 const double MAX_DOWNSAMPLE_FACTOR_SCALE{3.0};
-const std::size_t MAX_NUMBER_FOLDS{5};
+// This isn't a hard limit be we increase the number of default training folds
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeFactory.cc]: I think, something wrong with this sentence.

File: lib/maths/CBoostedTreeFactory.cc [hunk 8]
```diff
@@ -250,20 +253,32 @@ void CBoostedTreeFactory::initializeNumberFolds(core::CDataFrame& frame) const {
...
+        //   2. it still means we'll have plenty of variation between random bags.
+        //
+        // In order to estimate this we use the number of input features as a proxy
+        // for the number of features we'll actually use after feature selection.
+        //
+        // So how does the following work: we'd like "c * f * # rows" training rows.
+        // For k folds we'll have "(1 - 1 / k) * # rows" training rows. So we want
+        // to find the smallest integer k s.t. c * f * # rows <= (1 - 1 / k) * # rows.
+        // This gives k = ceil(1 / (1 - c * f)). However, we also upper bound this
+        // by MAX_NUMBER_FOLDS.
```
Comment (valeriy42) [inline lib/maths/CBoostedTreeFactory.cc]: This is a very nice explanation!
